ixmp4/data/db/optimization/indexset/model.py

```diff
@@ -10,26 +8,50 @@
 from .. import base
 
 
+# TODO Feels like there ought to be this kind of functionality already
+def cast_data_as_type(
+    data: "IndexSetData", type: Literal["float", "int", "str"] | None
+) -> float | int | str:
+    if type == "str":
+        return data.value
+    elif type == "int":
+        return int(data.value)
+    elif type == "float":
+        return float(data.value)
+    else:  # type is None
+        return 0
+
+
 class IndexSet(base.BaseModel):
     NotFound: ClassVar = abstract.IndexSet.NotFound
     NotUnique: ClassVar = abstract.IndexSet.NotUnique
     DataInvalid: ClassVar = OptimizationDataValidationError
     DeletionPrevented: ClassVar = abstract.IndexSet.DeletionPrevented
 
-    elements: types.JsonList = db.Column(db.JsonType, nullable=False, default=[])
+    data_type: types.OptimizationDataType
+
+    _data: types.Mapped[list["IndexSetData"]] = db.relationship(
+        back_populates="indexset"
+    )
 
-    @validates("elements")
-    def validate_elements(self, key, value: list[float | int | str]):
-        unique = set()
-        for element in value:
-            if element in unique:
-                raise self.DataInvalid(
-                    f"{element} already defined for IndexSet {self.name}!"
-                )
-            else:
-                unique.add(element)
-        return value
+    @db.hybrid_property
+    def data(self) -> list[float | int | str]:
+        return [cast_data_as_type(data, self.data_type) for data in self._data]
```

Comment:
This also looks quite slow, but is porbably unavoidable?

Reply:
Maybe @danielhuppmann can clarify, but I thought we want the `indexset.data` attribute to accurately portray the data type. So I guess we can either cast types here or in the API *and* core layers. 
But the function itself I was also wondering about: there should be some efficient built-in that does this kind of casting, my intuition tells me. Do you know of a better way? Would it be faster to collect all `self._data.value`s in a numpy array or so and cast that appropriately?

Reply:
According [to this guide](https://www.geeksforgeeks.org/python-type-casting-whole-list-and-matrix/), the `O()` of using loops and using numpy is both `O(n)`, but locally, the `tests/data` test runs 0.1 seconds faster when using numpy as it is used now, so hope that's better :)

Reply:
Well of course they are both O(n) yes, doesnt really say much about the actual cost though.

Reply:
I always thought it did, assuming the `n` can't be made smaller or so. Looking forward to your workshop on this :)